### PR TITLE
txsource: runtime workload

### PR DESCRIPTION
**Pull request by ptrus**
Runtime part of: #2506

TODO: 
- [x] validate runtime transaction results

### CODE REVIEW

File: tests/fixture-data/txsource/staking-genesis.json [hunk 1]
```diff
@@ -9,5 +9,20 @@
     },
     "fee_split_vote": "1",
     "fee_split_propose": "1"
+  },
+  "total_supply": "90000000",
+  "ledger": {
+    "LQu4ZtFg8OJ0MC4M4QMeUR7Is6Xt4A/CW+PK/7TPiH0=": {"general": {"balance": "10000000"}},
```
Comment (kostko) [inline tests/fixture-data/txsource/staking-genesis.json]: Make sure that this is enough for really long term tests (it's probably not).

Comment (ptrus) [reply tests/fixture-data/txsource/staking-genesis.json]: i've increased `fundAccountAmount` 100x, and each node account in genesis get's the same amount

File: go/oasis-node/cmd/debug/txsource/workload/runtime.go [hunk 2]
```diff
@@ -0,0 +1,230 @@
...
+
+	return nil
+}
+
+func (r *runtime) Run(gracefulExit context.Context, rng *rand.Rand, _ *grpc.ClientConn, _ consensus.ClientBackend, rtc runtimeClient.RuntimeClient, fundingAccount signature.Signer) error {
+	ctx := context.Background()
+
+	r.logger = logging.GetLogger("cmd/txsource/workload/runtime")
+	// Simple-keyvalue runtime.
+	_ = r.runtimeID.UnmarshalHex("8000000000000000000000000000000000000000000000000000000000000000")
```
Comment (kostko) [inline go/oasis-node/cmd/debug/txsource/workload/runtime.go]: Could this be changed to be a CLI argument instead of being hardcoded here? And then passed from txsource E2E test.

File: go/oasis-test-runner/scenario/e2e/txsource.go [hunk 3]
```diff
@@ -29,31 +29,34 @@ const (
 
 	nodeRestartIntervalLong = 2 * time.Minute
 	livenessCheckInterval   = 1 * time.Minute
+	txSourceGasPrice        = 1
 )
 
 // TxSourceMultiShort uses multiple workloads for a short time.
 var TxSourceMultiShort scenario.Scenario = &txSourceImpl{
-	basicImpl: *newBasicImpl("txsource-multi-short", "", nil),
+	basicImpl: *newBasicImpl("txsource-multi-short", "simple-keyvalue-client", nil),
```
Comment (kostko) [inline go/oasis-test-runner/scenario/e2e/txsource.go]: Why is this change needed? The scenario does not actually run this client? (Also below.)

Comment (ptrus) [reply go/oasis-test-runner/scenario/e2e/txsource.go]: thanks for noticing, leftover from some of my initial testing :)

File: go/oasis-node/cmd/debug/txsource/workload/runtime.go (hunk 2 above)
Comment (ptrus) [reply go/oasis-node/cmd/debug/txsource/workload/runtime.go]: :+1:

File: go/oasis-node/cmd/debug/txsource/workload/runtime.go [hunk 4]
```diff
@@ -0,0 +1,257 @@
...
+		r.logger.Error("runtime unmsrshal error",
+			"err", err,
+			"runtime_id", viper.GetString(CfgRuntimeID),
+		)
+		return fmt.Errorf("Runtime unmarshal: %w", err)
+	}
+	r.state = make(map[string]string)
+
+	for {
+		switch rng.Intn(3) {
```
Comment (pro-wh) [inline go/oasis-node/cmd/debug/txsource/workload/runtime.go]: if you're up for more fiddliness, make these weighted and configurable too

File: go/oasis-node/cmd/debug/txsource/workload/runtime.go [hunk 5]
```diff
@@ -0,0 +1,257 @@
...
+		case 0:
+			if err := r.doInsertRequest(ctx, rng, rtc, rng.Float64() < runtimeInsertExistingRatio); err != nil {
+				return err
```
Comment (pro-wh) [inline go/oasis-node/cmd/debug/txsource/workload/runtime.go]: no context added, that it's an insert request

File: go/oasis-node/cmd/debug/txsource/workload/runtime.go [hunk 6]
```diff
@@ -0,0 +1,257 @@
...
+	})
+	if err != nil {
+		return fmt.Errorf("submit remove request failed: %w", err)
+	}
+	r.logger.Debug("remove request success",
+		"response", rsp,
+	)
+
+	if err := r.validateResponse(key, rsp); err != nil {
+		return fmt.Errorf("invalid response: %w", err)
```
Comment (pro-wh) [inline go/oasis-node/cmd/debug/txsource/workload/runtime.go]: some of these are missing context info:
- what kind of request it was
- info about the request (should we?)

File: go/oasis-test-runner/scenario/e2e/txsource.go [hunk 7]
```diff
@@ -219,6 +237,7 @@ func (sc *txSourceImpl) startWorkload(childEnv *env.Env, errCh chan error, name
 		"--log.format", logFmt.String(),
 		"--log.level", logLevel.String(),
 		"--" + flags.CfgGenesisFile, sc.net.GenesisPath(),
+		"--" + workload.CfgRuntimeID, runtimeID.String(),
```
Comment (pro-wh) [inline go/oasis-test-runner/scenario/e2e/txsource.go]: :information_source: `runtimeID` is from common.go, and it's the zero value

File: go/oasis-test-runner/scenario/e2e/txsource.go [hunk 8]
```diff
@@ -76,17 +79,32 @@ func (sc *txSourceImpl) Fixture() (*oasis.NetworkFixture, error) {
...
 	f.Network.StakingGenesis = "tests/fixture-data/txsource/staking-genesis.json"
 
 	// Disable CheckTx on the client node so we can submit invalid transactions.
 	f.Clients[0].ConsensusDisableCheckTx = true
 
-	// MinGasPrice
-	f.Validators = []oasis.ValidatorFixture{
-		oasis.ValidatorFixture{Entity: 1, MinGasPrice: 1},
-		oasis.ValidatorFixture{Entity: 1, MinGasPrice: 1},
-		oasis.ValidatorFixture{Entity: 1, MinGasPrice: 1},
```
Comment (pro-wh) [inline go/oasis-test-runner/scenario/e2e/txsource.go]: how did this work before, was `SubmissionGasPrice` by default 1?

File: tests/fixture-data/txsource/staking-genesis.json [hunk 9]
```diff
@@ -9,5 +9,20 @@
...
+    "toAWSrxsxSFbGXQIhu4p2fOL0z/fnirrGT/MuzRgcW8=": {"general": {"balance": "10000000000"}},
+
+    "DbeoxcRwDO4Wh8bwq5rAR7wzhiB+LeYn+y7lFSGAZ7I=": {"general": {"balance": "10000000000"}},
+    "oWk0qdNhp7XNQDX4YFJY2KeFM+9u4fo16DeSdW1gChg=": {"general": {"balance": "10000000000"}},
+    "hcWVUucu/MwOyEDmj5T4TACwpgre5e2cJLczBggxxm0=": {"general": {"balance": "10000000000"}},
+
+    "B14NG+Joea3j09rVt0SEPQ4Z6Ned1EEi6rzZoa2YrYw=": {"general": {"balance": "10000000000"}},
+    "iPYyR3I45FRGK568PTJIZUmELgcFYPhIb80e2GxrIf0=": {"general": {"balance": "10000000000"}},
+
+    "mh+b9hN+2oLAflZACgBgVHA0EjeZK3ew6kml75s2wN4=": {"general": {"balance": "10000000000"}}
```
Comment (pro-wh) [inline tests/fixture-data/txsource/staking-genesis.json]: whose keys are these?

why do they need these funds now?

File: go/oasis-node/cmd/debug/txsource/workload/runtime.go [hunk 10]
```diff
@@ -0,0 +1,257 @@
...
+	state     map[string]string
+}
+
+func (r *runtime) generateVal(rng *rand.Rand, existingKey bool) string {
+	if existingKey && len(r.state) > 0 {
+		// Select existing key to be used.
+		keyIdx := rng.Intn(len(r.state))
+		i := 0
+		for k := range r.state {
+			if i == keyIdx {
```
Comment (pro-wh) [inline go/oasis-node/cmd/debug/txsource/workload/runtime.go]: aw man linear scan

File: go/oasis-node/cmd/debug/txsource/txsource.go [hunk 11]
```diff
@@ -28,7 +28,7 @@ import (
 const (
 	CfgWorkload  = "workload"
 	CfgSeed      = "seed"
-	CfgTimeLimit = "time_limit"
+	CfgTimeLimit = "time-limit"
```
[the remark this answers is not in the document]
Comment (ptrus) [reply go/oasis-node/cmd/debug/txsource/txsource.go]: ah no, i was misguided by the `--runtime-id`  & `--node-address` flags. seems like for most others we use `_` as a separator, will change back.

File: go/oasis-node/cmd/debug/txsource/workload/delegation.go [hunk 12]
```diff
@@ -229,7 +229,14 @@ func (d *delegation) doReclaimEscrowTx(ctx context.Context, rng *rand.Rand, cnsc
...
 
-func (d *delegation) Run(gracefulExit context.Context, rng *rand.Rand, conn *grpc.ClientConn, cnsc consensus.ClientBackend, rtc runtimeClient.RuntimeClient, fundingAccount signature.Signer) error {
+func (d *delegation) Run(
+	gracefulExit context.Context,
+	rng *rand.Rand,
+	conn *grpc.ClientConn,
+	cnsc consensus.ClientBackend,
+	rtc runtimeClient.RuntimeClient,
+	fundingAccount signature.Signer,
+) error {
```
[the remark this answers is not in the document]
Comment (ptrus) [reply go/oasis-node/cmd/debug/txsource/workload/delegation.go]: yeah

File: go/oasis-node/cmd/debug/txsource/workload/runtime.go [hunk 13]
```diff
@@ -0,0 +1,257 @@
...
+	"github.com/oasislabs/oasis-core/go/common/cbor"
+	"github.com/oasislabs/oasis-core/go/common/crypto/signature"
+	"github.com/oasislabs/oasis-core/go/common/logging"
+	consensus "github.com/oasislabs/oasis-core/go/consensus/api"
+	runtimeClient "github.com/oasislabs/oasis-core/go/runtime/client/api"
+	runtimeTransaction "github.com/oasislabs/oasis-core/go/runtime/transaction"
+)
+
+const (
+	// NameRuntime is the name of the Runtime workload.
```
[the remark this answers is not in the document]
Comment (ptrus) [reply go/oasis-node/cmd/debug/txsource/workload/runtime.go]: fixed

File: go/oasis-test-runner/scenario/e2e/txsource.go (hunk 8 above)
Comment (ptrus) [reply go/oasis-test-runner/scenario/e2e/txsource.go]: validators accepted their own tx'es (no fee's required), other nodes failed registrations but none of the workload tests tested anything runtime related so no test failed :-)

File: tests/fixture-data/txsource/staking-genesis.json (hunk 9 above)
Comment (ptrus) [reply tests/fixture-data/txsource/staking-genesis.json]: these are from 
```
// Use deterministic identities as we need to allocate funds to nodes.
f.Network.DeterministicIdentities = true
```

Before these PR only validators successfully registered see

File: go/oasis-test-runner/scenario/e2e/txsource.go (hunk 7 above)
Comment (ptrus) [reply go/oasis-test-runner/scenario/e2e/txsource.go]: it's not the zero value: https://github.com/oasislabs/oasis-core/blob/master/go/oasis-test-runner/scenario/e2e/common.go#L115

File: go/oasis-node/cmd/debug/txsource/workload/runtime.go [hunk 14]
```diff
@@ -0,0 +1,257 @@
...
+		// expected.
+		if rsp.Success != nil {
+			return fmt.Errorf("expected nil response, got: '%s'", rsp.Success)
+		}
+	}
+
+	return nil
+}
+
+func (r *runtime) submitRuntimeRquest(ctx context.Context, rtc runtimeClient.RuntimeClient, req interface{}) (*runtimeTransaction.TxnOutput, error) {
```
[the remark this answers is not in the document]
Comment (ptrus) [reply go/oasis-node/cmd/debug/txsource/workload/runtime.go]: :+1: updated

File: go/oasis-node/cmd/debug/txsource/workload/runtime.go (hunk 5 above)
Comment (ptrus) [reply go/oasis-node/cmd/debug/txsource/workload/runtime.go]: fixed, also bellow

File: go/oasis-node/cmd/debug/txsource/workload/runtime.go [hunk 15]
```diff
@@ -0,0 +1,257 @@
...
+)
+
+// RuntimeFlags are the runtime workload flags.
+var RuntimeFlags = flag.NewFlagSet("", flag.ContinueOnError)
+
+type runtime struct {
+	logger *logging.Logger
+
+	runtimeID common.Namespace
+	state     map[string]string
```
[the remark this answers is not in the document]
Comment (ptrus) [reply go/oasis-node/cmd/debug/txsource/workload/runtime.go]: renamed to `reckonedKeyValueState`

File: go/oasis-node/cmd/debug/txsource/workload/runtime.go (hunk 6 above)
Comment (ptrus) [reply go/oasis-node/cmd/debug/txsource/workload/runtime.go]: Fixed, also elsewhere. Added log messages that log the request, response, error.

File: go/oasis-node/cmd/debug/txsource/workload/runtime.go (hunk 4 above)
Comment (ptrus) [reply go/oasis-node/cmd/debug/txsource/workload/runtime.go]: done :+1:

**Comment by ptrus** [comment]
tanks for the review comments @pro-wh please check again

File: go/oasis-node/cmd/debug/txsource/txsource.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (pro-wh) [reply go/oasis-node/cmd/debug/txsource/txsource.go]: 🤷 ok

File: go/oasis-test-runner/scenario/e2e/txsource.go (hunk 7 above)
Comment (pro-wh) [reply go/oasis-test-runner/scenario/e2e/txsource.go]: ah thanks for the correction

File: go/oasis-test-runner/scenario/e2e/txsource.go (hunk 8 above)
Comment (pro-wh) [reply go/oasis-test-runner/scenario/e2e/txsource.go]: ok

File: tests/fixture-data/txsource/staking-genesis.json (hunk 9 above)
Comment (pro-wh) [reply tests/fixture-data/txsource/staking-genesis.json]: ok thanks

File: go/oasis-node/cmd/debug/txsource/workload/runtime.go [hunk 16]
```diff
@@ -0,0 +1,311 @@
...
+		}{
+			Key:   key,
+			Value: value,
+		},
+	}
+	rsp, err := r.submitRuntimeRquest(ctx, rtc, req)
+	if err != nil {
+		r.logger.Error("Submit insert request failure",
+			"request", req,
+			"response", rsp,
```
Comment (pro-wh) [inline go/oasis-node/cmd/debug/txsource/workload/runtime.go]: I think it's not valid to access `rsp` in the err case
